apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/reporting/ReportingTracer.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.tracer.reporting;
+
+import co.elastic.apm.agent.tracer.Tracer;
+
+import java.io.Closeable;
+import java.util.concurrent.ScheduledExecutorService;
+
+public interface ReportingTracer extends Tracer {
+
+    void addMetric(String name, Labels labels, DoubleSupplier metric);
+
+    void removeMetric(String name, Labels labels);
+
+    /**
+     * Reports an ECS-logging formatted log message.
+     *
+     * @param log log message generated by ecs-logging
+     */
+    void reportLog(String log);
+
+    /**
+     * Reports an ECS-logging formatted log message.
+     *
+     * @param log log message generated by ecs-logging
+     */
+    void reportLog(byte[] log);
+
+    DataWriter newWriter(int maxSerializedSize);
+
+    ScheduledExecutorService getSharedSingleThreadedPool();
+
+    void addShutdownHook(Closeable job);
```

Comment:
Wouldn't it make more sense to have shutdown hooks as part of `LifecycleListener`s?

Reply:
I thought about it myself but the hook is registered programmatically and the instance state cannot be transferred that easily.

Reply:
Ah, I see. The problem is that `LifecycleListener`s are instantiated as singletons, and the singletons are not accessible from instrumentations without piping them through an ugly `static` variable.

In my opinion this should rater be solved by making `ElasticApmTracer.getLifecycleListener` part of the tracer-api.
This way instrumentations also get access to the `start()` method, which could be called after `stop()` if the tracer is resumed.

I had a look at the only usage of `addShutdownHook`, which currently is the micrometer instrumentation. This instrumentation would then register a custom `LifecycleListener`, which in turn allows the `MicrometerMetricsReporter`s to register themselves to be flushed on shutdown.

Reply:
Ok, after an internal discussion we decided to stay with just adding `addShutdownHook` to the `Tracer`. This should keep things simple for now, if we want to, we can refactor later.